src/core/mesh/qgsmeshvectorrenderer.cpp

```diff
@@ -0,0 +1,264 @@
+  mMaxX = *std::max_element( datasetValuesX.constBegin(), datasetValuesX.constEnd() );
+  mMinY = *std::min_element( datasetValuesY.constBegin(), datasetValuesY.constEnd() );
+  mMaxY = *std::max_element( datasetValuesY.constBegin(), datasetValuesY.constEnd() );
+  mMinMag = *std::min_element( mDatasetValuesMag.constBegin(), mDatasetValuesMag.constEnd() );
+  mMaxMag = *std::max_element( mDatasetValuesMag.constBegin(), mDatasetValuesMag.constEnd() );
+}
+
+QgsMeshVectorRenderer::~QgsMeshVectorRenderer() = default;
+
+void QgsMeshVectorRenderer::draw()
+{
+  // Set up the render configuration options
+  QPainter *p = mContext.painter();
+  p->setRenderHint( QPainter::Antialiasing );
+  QPen pen = p->pen();
+  pen.setCapStyle( Qt::FlatCap );
+  pen.setJoinStyle( Qt::MiterJoin );
+
+  double penWidth = mContext.convertToPainterUnits( mCfg.lineWidth(),
+                    QgsUnitTypes::RenderUnit::RenderMillimeters );
+  pen.setWidthF( penWidth );
+  pen.setColor( mCfg.color() );
+  p->setPen( pen );
+
+  if ( mDataOnVertices )
+    drawVectorDataOnVertices();
+  else
+    drawVectorDataOnFaces();
+}
+
+bool QgsMeshVectorRenderer::calcVectorLineEnd(
+  QgsPointXY &lineEnd,
+  double &vectorLength,
+  double &cosAlpha,
+  double &sinAlpha, //out
+  const QgsPointXY &lineStart,
+  double xVal,
+  double yVal,
+  double magnitude //in
+)
+{
+  // return true on error
+
+  if ( xVal == 0.0 && yVal == 0.0 )
+    return true;
+
+  // do not render if magnitude is outside of the filtered range (if filtering is enabled)
+  if ( mCfg.filterMin() >= 0 && magnitude < mCfg.filterMin() )
+    return true;
+  if ( mCfg.filterMax() >= 0 && magnitude > mCfg.filterMax() )
+    return true;
+
+  // Determine the angle of the vector, counter-clockwise, from east
+  // (and associated trigs)
+  double vectorAngle = -1.0 * atan( ( -1.0 * yVal ) / xVal );
+  cosAlpha = cos( vectorAngle ) * mag( xVal );
+  sinAlpha = sin( vectorAngle ) * mag( xVal );
+
+  // Now determine the X and Y distances of the end of the line from the start
+  double xDist = 0.0;
+  double yDist = 0.0;
+  switch ( mCfg.shaftLengthMethod() )
+  {
+    case QgsMeshRendererVectorSettings::VectorRenderingType::MinMax:
+    {
+      double minShaftLength = mContext.convertToPainterUnits( mCfg.minShaftLength(),
+                              QgsUnitTypes::RenderUnit::RenderMillimeters );
+      double maxShaftLength = mContext.convertToPainterUnits( mCfg.maxShaftLength(),
+                              QgsUnitTypes::RenderUnit::RenderMillimeters );
+      double minVal = mMinMag;
+      double maxVal = mMaxMag;
+      double k = ( magnitude - minVal ) / ( maxVal - minVal );
+      double L = minShaftLength + k * ( maxShaftLength - minShaftLength );
+      xDist = cosAlpha * L;
+      yDist = sinAlpha * L;
+      break;
+    }
+    case QgsMeshRendererVectorSettings::VectorRenderingType::Scaled:
+    {
+      double scaleFactor = mCfg.scaleFactor();
```

Comment:
I think here we should also convert from millimeters to pixels? otherwise the scaling will be to pixels...

Reply:
don't think so, this is just a factor to multiply the magnitude, no units